### ISSUE TITLE
ClickHouse and ML

### ISSUE THREAD

**Issue by KochetovNicolai** [ClickHouse/ClickHouse#7345]
There are some features connected with machine learning which are good to implement in ClickHouse. Descriptions are approximate, not necessary to follow then completely. It is also ok to implement something partially.

### Loss functions (MSE, MAE, logloss, ...)
Most of loss functions could be calculated manually, using standard SQL. Example for MSE:

```{sql}
SELECT sqrt(avg(diff * diff)) AS MSE from ...
```

where `diff` is a difference between target and prediction. 
 
Still, it would be convenient to have specialised loss functions. MSE could have the following interface:
```{sql}
SELECT MSE(target, prediction) from ...
```
 
 In the example above `MSE` is an **aggregate function**. You can read an implementation of `avg` function as an example. 
 
##### Implementation details to consider
* Function arguments should support any numeric type. Inner implementation may cast arguments to `Float64`.
* Try to avoid numeric overflow, zero division and nan's. Also consider detection of outliers: when `abs(target - prediction)` is much more than average value.
* It would be good if you add support for `Nullable` types. You can skip them. 
* Think about common interface for all loss functions, especially if you want to implement several ones.

### Shuffle for minibatches (as table function)
Minibatch shuffling allows to achieve better quality for models based on gradient descent. 
It is possible to implement shuffling as table function with interface:

```
SELECT ... from suffle(database, table, minibatch_size)
```
or
```
SELECT ... from suffle(database.table, minibatch_size)
```
or
```
SELECT ... from suffle(table_function, minibatch_size)
```
Data in ClickHouse functions are processed by blocks (see developer instruction for details). Table function `shuffle` should return block with shuffled rows.
In the examples above `minibatch_size` is an optional argument. If it is not specified, blocks size shouldn't changed.
Otherwise, result block should have a size of `minibatch_size`.

### Table function to sample data from table with repetition
While training of a model, dataset could be repeated several times. 
In common case, when all data could be placed in memory, it could be implemented as mere sampling of random rows
 (evenly distributed, with rows repetition). Similar result could be achieved in ClikHouse using minibatch shuffling and 
 repeated reading from table. 
 
 It is possible, but not convenient, to read from table several times using `UNION ALL` keyword form `SQL`. 
 However, better approach is to implement specialized table functions with following interface:
 ```SQL
SELECT ... from repeat(database, table, num_iterations)
``` 
and
```sql
SELECT ... from loop(database, table, limit)
```
Function `repeat` should read table `num_iteration` times. Function `loop` should read table in infinite loop until `limit` rows is achieved (is specified).

You can implement other interfaces from `shuffle` example.

### Support for multiple features in simpleLinearRegression

Possible interface:
```SQL
SELECT simpleLiniarRegression(x1, ..., xN, y)
```
or
```SQL
SELECT simpleLiniarRegression(tuple(x1, ..., xN), y)
```

See `AggregateFunctionSimpleLinearRegression.h` for details.

### Support for multiple loss functions in stochasticLinearRegression
Now `stochasticLinearRegression` is implemented for `MSE` loss function. You can implement support for other loss functions and add loss function name as another parameter.

Consider of creating new interface for loss functions (it is different from interface for the same aggregate functions from the first item).
It should allow calculation of loss function's gradients (probably, for batch execution also).

See `AggregateFunctionMLMethod.h` for details.

### Named parameters for aggregate functions
Aggregate functions in ClickHouse now only support positional parameters. So, if I want to change only optimizer for function `stochasticLinearRegression`, I need to specify all the parameters. For example:
```{sql}
SELECT stochasticLinearRegression(0.1, 1, 32, 'Nesterov')(y, x1, x2) FROM ...
```
It is also esy to mix arguments with the same types.

It would be good to add support for named parameters. Example:
```{sql}
SELECT stochasticLinearRegression(optimizer='Nesterov', learning_rate=0.1)(y, x1, x2) FROM ...
```
All parameters which are not specified are considered as default. 
For simplicity, you can assume that all arguments must either named or positional.

### More aggregate functions for ML
You are welcome to implement any ML method you want. It can be done as an aggregate function (is done for regression methods) or as an external model (is done for CatBoost).
It is perfectly ok to support only applying of already trained model.

Please, share for your ideas in comments.

**Comment by alex-krash** [comment]
@KochetovNicolai , I may implement loss functions (we are currently using them for timeseries analysis already, and I was thinking that they should be implemented as part of CH). Should I create a separate issue for discussion of it?

**Comment by KochetovNicolai** [comment]
@alex-krash that's great! You may create separate issue if you want. Or just create a pr :)

**Comment by alexey-milovidov** [comment]
Nothing was done. Closing.

**Comment by alexey-milovidov** [comment]
Even the half-ready PR with data clustering was abandoned.

**Comment by KochetovNicolai** [comment]
RIP
